truncate/table_schema.go

```diff
@@ -137,6 +131,130 @@ func fetchTableSchemas(ctx context.Context, client *spanner.Client, targetTables
+	}
+
+	// TODO: Add child tables that may be deleted in cascade (#18)
+
+	filtered := make([]*tableSchema, 0, len(tables))
+	for _, t := range tables {
+		if isTarget[t.tableName] {
+			filtered = append(filtered, t)
+		}
+	}
+
+	return filtered
+}
+
+// excludeFilterTableSchemas filters tables with given excludeTables.
+// If excludeTables is empty, it returns all tables.
+// When an exclude table is cascade deletable, its parent table is also excluded.
+func excludeFilterTableSchemas(tables []*tableSchema, excludeTableSchemas []string) []*tableSchema {
+	if len(excludeTableSchemas) == 0 {
+		return tables
+	}
+
+	isExclude := make(map[string]bool, len(tables))
+	for _, t := range excludeTableSchemas {
+		isExclude[t] = true
+	}
+
+	// Additionally exclude parent tables that may delete the exclude tables in cascade
+	lineages := constructTableLineages(tables)
+	for _, l := range lineages {
+		if isExclude[l.tableSchema.tableName] {
+			for _, a := range l.ancestors {
+				isExclude[a.tableName] = true
+			}
+		}
+	}
+
+	filtered := make([]*tableSchema, 0, len(tables))
+	for _, t := range tables {
+		if !isExclude[t.tableName] {
+			filtered = append(filtered, t)
+		}
+	}
+
+	return filtered
+}
+
+// constructTableLineages returns a list of interleave Lineages.
+// This function creates tableLineage for each of all given tableSchemas.
+func constructTableLineages(tables []*tableSchema) []*tableLineage {
+	tableMap := make(map[string]*tableSchema, len(tables))
+	for _, t := range tables {
+		tableMap[t.tableName] = t
+	}
+
+	parentRelation := make(map[string]*tableSchema, len(tables))
+	childRelation := make(map[string][]*tableSchema, len(tables))
+	for _, t := range tables {
+		if !t.isRoot() && t.isCascadeDeletable() {
+			parentRelation[t.tableName] = tableMap[t.parentTableName]
+			childRelation[t.parentTableName] = append(childRelation[t.parentTableName], t)
+		}
+	}
+
+	lineages := make([]*tableLineage, 0, len(tables))
+	for _, t := range tables {
+		lineages = append(lineages, &tableLineage{
+			tableSchema: t,
+			ancestors:   findAncestors(t, parentRelation, nil),
+			descendants: findDescendants(t, childRelation, nil),
+		})
+	}
+
+	return lineages
+}
+
+// findAncestors recursively finds all ancestors of the given table .
+func findAncestors(table *tableSchema, parentRelation map[string]*tableSchema, ancestors []*tableSchema) []*tableSchema {
+	if parent, ok := parentRelation[table.tableName]; ok {
+		ancestors = append(ancestors, parent)
+		return findAncestors(parent, parentRelation, ancestors)
+	}
+	return ancestors
+}
+
+// findDescendants recursively finds all descendants of the given table.
+func findDescendants(table *tableSchema, childRelation map[string][]*tableSchema, descendants []*tableSchema) []*tableSchema {
+	for _, child := range childRelation[table.tableName] {
+		descendants = append(descendants, child)
+		descendants = findDescendants(child, childRelation, descendants)
+	}
+	return descendants
+}
```

Comment:
Implemented hierarchical traversal using recursion, resulting in relatively readable code. Thank you for your advice!